src/text.rs

```diff
@@ -28,6 +30,25 @@ impl Text {
 
 impl AsRef<CStr> for Text {
     fn as_ref(&self) -> &CStr {
+        if self.0.is_null() {
+            // TODO: This is a dumb hack.
+            // Tesseract may choose to return a null pointer for no text for some invalid states.
```

Comment:
I suppose you've done it this way to avoid a breaking change.

But if we were to release a breaking change, do you see this case being:
1. `Option::None`
2. `Result::Err`
3. Or keep it as an empty string?

Reply:
I think a fallible constructor or simply noting the safety requirements on the constructor may be fine, but the current upstream crates make the assumption that it's safe to deref as a cstr so I considered this fixing a safety issue in the current version.

The instances I encountered `Text`s being null were as a result of errors I encountered as setup/config of the image iteration so I'd honestly be more inclined to believe it's an `Err` than anything for the relevant get_text() functions purely on the basis of naming but I'm not sure if there's any other error states that can be encoded other than `null`, so I can see an argument for either.

Reply:
What about doing something like this to enforce the pointer isn't `null` when `Text` is initialized?

a0926633130fb4e0f7aef1a3a8e1a5f7834f8f4c

It's technically a breaking change.

Reply:
Aside from the (IMO) far too wordy `TessBaseApiGetWordStrBoxTextError` naming thing,

 `impl Into<TessBaseApiGetWordStrBoxTextError> for TextNewError` for the associated error types  allows for reducing all the `Text` constructors to
```rust
pub fn get_utf8_text(&mut self) -> Result<Text, TessBaseApiGetUtf8TextError> {
   let ptr = unsafe { TessBaseAPIGetUTF8Text(self.0) }; // Or similar calls
   Text::new(ptr)?
}
```

Reply:
> Aside from the (IMO) far too wordy `TessBaseApiGetWordStrBoxTextError` naming thing (I don't know what `StrBox` is referring to in this situ)

That's basically the error from the [TessBaseAPIGetWordStrBoxText](https://tesseract-ocr.github.io/tessapi/5.x/a00008.html#ab9938845c9b52434ee32a2225aad81cf) method.

There is more documentation about that [here](https://tesseract-ocr.github.io/tessapi/5.x/a02438.html#aa45bc5ee67de0bf66112165ca93399ff). I don't know what a `StrBox` is either, but it probably makes sense to the original c++ developers.

Reply:
`Box` might be referring to a rectangle within the image, rather than heap memory like in Rust.

Reply:
> `impl Into<TessBaseApiGetWordStrBoxTextError> for TextNewError` for the associated error types allows for reducing all the `Text` constructors to
> 
> ```rust
> pub fn get_utf8_text(&mut self) -> Result<Text, TessBaseApiGetUtf8TextError> {
>    let ptr = unsafe { TessBaseAPIGetUTF8Text(self.0) }; // Or similar calls
>    Text::new(ptr)?
> }
> ```

```rust
    pub fn get_utf8_text(&mut self) -> Result<Text, TextNewError> {
        unsafe { Text::new(TessBaseAPIGetUTF8Text(self.0)) }
    }
```

What about this, and just give up on the custom errors? In hindsight there isn't much point having specific errors for all the text functions as they all have the one shared error case (null pointer).

Reply:
I'd be for eliminating function-specific errors in this situation, I'm not sure if we can actually provide more "useful" error messages with the API calls at this level. 
Ideally, I'd like to be able to document *why* errors could occur as there's definitely a level of "order of operations" to the API I feel is undocumented/unclear within the C API but I'm also not very fluent in C/C++.

Reply:
I've made the change to stick to a single error for the text conversions here: 21ae92032e489222310722ff8ca60e18f5d7ccc1

I have a branch that includes both our changes https://github.com/ccouzens/tesseract-plumbing/tree/StripedMonkey-main



> I'm not sure if we can actually provide more "useful" error messages with the API calls at this level.

Yeah, I think at this point all we can say is we received a `null` pointer.

> Ideally, I'd like to be able to document _why_ errors could occur as there's definitely a level of "order of operations" to the API I feel is undocumented/unclear within the C API but I'm also not very fluent in C/C++.

Yeah, for example that you actually set an image. The higher level crates that build off of here (`leptess` and `tessearct-rs`), both do that to a degree.